src/lib/keccak.ts

```diff
@@ -209,18 +389,233 @@ function permutation(state: Field[][], rc: Field[]): Field[][] {
+  const squeezes = Math.floor(length / rate) + 1;
+  // multiple of rate that is larger than output_length, in bytes
+  const outputLength = squeezes * bytesPerSqueeze;
+  // array with sufficient space to store the output
+  const outputArray = Array(outputLength).fill(Field.from(0));
+  // first state to be squeezed
+  const bytestring = keccakStateToBytes(state);
+  const outputBytes = bytestring.slice(0, bytesPerSqueeze);
+  copy(outputBytes, outputArray, 0, bytesPerSqueeze);
+
+  // for the rest of squeezes
+  for (let i = 1; i < squeezes; i++) {
+    // apply the permutation function to the state
+    newState = permutation(newState, rc);
+    // append the output of the permutation function to the output
+    const bytestringI = keccakStateToBytes(state);
+    const outputBytesI = bytestringI.slice(0, bytesPerSqueeze);
+    copy(outputBytesI, outputArray, bytesPerSqueeze * i, bytesPerSqueeze);
+  }
+
+  // Obtain the hash selecting the first bitlength/8 bytes of the output array
+  const hashed = outputArray.slice(0, length / 8);
+  return hashed;
+}
+
+// Keccak sponge function for 1600 bits of state width
+// Need to split the message into blocks of 1088 bits.
+function sponge(
+  paddedMessage: Field[],
+  length: number,
+  capacity: number,
+  rate: number
+): Field[] {
+  // check that the padded message is a multiple of rate
+  if ((paddedMessage.length * 8) % rate !== 0) {
+    throw new Error('Invalid padded message length');
+  }
+
+  // load round constants into Fields
+  let rc = exists(24, () => TupleN.fromArray(24, ROUND_CONSTANTS));
+
+  // absorb
+  const state = absorb(paddedMessage, capacity, rate, rc);
+
+  // squeeze
+  const hashed = squeeze(state, length, rate, rc);
+
+  return hashed;
+}
+
+// TODO(jackryanservia): Use lookup argument once issue is resolved
+// Checks in the circuit that a list of Fields are at most 8 bits each
+function checkBytes(inputs: Field[]): void {
+  inputs.map(rangeCheck8);
+}
+
+// Keccak hash function with input message passed as list of Field bytes.
+// The message will be parsed as follows:
+// - the first byte of the message will be the least significant byte of the first word of the state (A[0][0])
+// - the 10*1 pad will take place after the message, until reaching the bit length rate.
+// - then, {0} pad will take place to finish the 1600 bits of the state.
+function hash(
+  inpEndian: 'Big' | 'Little' = 'Big',
+  outEndian: 'Big' | 'Little' = 'Big',
+  byteChecks: boolean = false,
+  message: Field[] = [],
+  length: number,
+  capacity: number,
+  nistVersion: boolean
+): Field[] {
+  assert(capacity > 0, 'capacity must be positive');
+  assert(
+    capacity < KECCAK_STATE_LENGTH,
+    'capacity must be less than KECCAK_STATE_LENGTH'
+  );
+  assert(length > 0, 'length must be positive');
+  assert(length % 8 === 0, 'length must be a multiple of 8');
+
+  // Input endianness conversion
+  let messageFormatted = inpEndian === 'Big' ? message : message.reverse();
```

Comment:
`message.reverse()` mutates the input, which no function should ever do unexpectedly. reverse a copy instead:
```suggestion
  let messageFormatted = inpEndian === 'Big' ? message : [...message].reverse();
```